base/compiler/typeinfer.jl

```diff
@@ -117,7 +117,7 @@ function cache_result(result::InferenceResult, min_valid::UInt, max_valid::UInt)
             end
             if !toplevel && inferred_result isa CodeInfo
                 cache_the_tree = result.src.inferred &&
-                    (result.src.inlineable ||
+                    (result.src.inlineable || result.src.pure ||
```

Comment:
You can't reliably assume that the contents of the `inferred` field will stick around. If you want to expose a property, it's probably best to make a boolean flag in the CodeInstance itself.